src/core/SampleLoader.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * SampleLoader.cpp - Sample loader with an optional cache
+ *
+ * Copyright (c) 2023 saker <sakertooth@gmail.com>
+ *               2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleLoader.h"
+
+#include <QMessageBox>
+#include <cassert>
+
+#include "Engine.h"
+#include "GuiApplication.h"
+#include "PathUtil.h"
+#include "Song.h"
+#include "lmms_basics.h"
+
+namespace lmms {
+
+namespace {
+
+void displayError(const QString& message)
+{
+	Engine::getSong()->collectError(QString("%1: %2").arg(QObject::tr("Error loading sample"), message));
+}
+
+} // namespace
+
+/**
+ * An evictor which automatically evicts sample buffers from the cache
+ * immediately after the last reference to the buffer is dropped
+*/
+class SampleLoader::AutoEvictor
+{
+public:
+	void operator()(SampleBuffer* p) const noexcept
+	{
+		if (!p) { return; }
+
+		SampleLoader::inst().remove(*p);
+		delete p;
+	}
+};
+
+SampleLoader::SampleLoader()
+{
+	connect(&m_watcher, &QFileSystemWatcher::fileChanged, this, &SampleLoader::removeFile);
+}
+
+auto SampleLoader::inst() -> SampleLoader&
+{
+	static SampleLoader singleton;
+	return singleton;
+}
+
+auto SampleLoader::fromFile(const QString& filePath, Cache cache)
+	-> std::shared_ptr<const SampleBuffer>
+{
+	if (filePath.isEmpty()) { return nullptr; }
+
+	switch (cache)
+	{
+		case Cache::Read: [[fallthrough]];
+		case Cache::ReadWrite:
+		{
+			if (auto buffer = SampleLoader::inst().getFile(filePath))
+			{
+				return buffer;
+			}
+		}
+		[[fallthrough]];
+		case Cache::None:
+		{
+			try
+			{
+				if (cache == Cache::ReadWrite)
+				{
+					auto buffer = std::shared_ptr<SampleBuffer> {
+						new SampleBuffer{SampleBuffer::Access{}, filePath},
+						AutoEvictor{}
+					};
+					SampleLoader::inst().add(buffer);
+					return buffer;
+				}
+
+				return std::make_shared<SampleBuffer>(SampleBuffer::Access{}, filePath);
+			}
+			catch (const std::runtime_error& error)
+			{
+				displayError(QString::fromStdString(error.what()));
```

Comment:
What I really don't like here is the exception handling (yes, I wrote this, but I actually think what I did here is not ideal).

A problem could occur for example when `SampleClip::loadSettings` gets a buffer from `SampleLoader`. When we load a project, we should be collecting errors using `Engine::getSong()->collectError`, not displaying a message box. Because the failure to load a buffer is handled differently, I think callers should be responsible for loading a buffer, or at least have the exception handling outside of `SampleLoader`.

Better yet, have exceptions from the failure to load a buffer be caught in `Song::loadProject`. I was in the process of implementing this but I never finished it because I wanted it to be done in a separate PR, so I added a TODO comment explaining that the exception handling can be improved in this way.

Reply:
> When we load a project, we should be collecting errors using Song::collectError, not displaying a message box.

`displayError()` calls `Song::collectError()` though, so we should be fine?

Reply:
Oh, didn't see that. I'm guessing you changed it here and I missed it. However, do we always want to collect errors? I think its more valuable in certain cases to display a message box to the user, so theres still this difference in error handling that is situational.

Reply:
By collecting errors when loading a project, it will display any missing samples or other issues to the user in a message box once the project finishes loading (see: `Song::loadProject()`), though I don't think it would do anything if an error occurred *after* the project was loaded - for example, if there's not enough memory to load a large sample.

Reply:
> though I don't think it would do anything if an error occurred after the project was loaded - for example, if there's not enough memory to load a large sample.

The main error could be that the user loads a sample from a dialog and the file is corrupted somehow, or the codec isn't supported, so we definitely should be handling these as a message box. I believe `Song::collectError` only applies when we load the project, so I dont think we could use it in all cases. I may be wrong here though.